onnxruntime/core/optimizer/reshape_fusion.cc

```diff
@@ -99,10 +98,6 @@ bool ReshapeFusion::Fuse_Subgraph1(Node& reshape, Graph& graph, const logging::L
     return false;
   }
 
-  if (graph_utils::GetInputNode(shape_1, 0) != p_root) {
```

Comment:
I have concern on loosing the constraints. The reason is the dim value 0 and -1 for Reshape are special, they shall be limited to the same tensor.

For example,
| Formula |                                                             Value|
|:--- |:--- | 
|s1 = Shape(x) |                                                      (2, 3, 4, 5)|
|d1 = s1[0]|                2|
|d2 = s2[1]|               3|
|s2 = [d1, d2, 20]|                                     [2, 3, 20]|
|z1 = Reshape(x, s2)| z has shape (2, 3, 20)|

We can optimize it as:
s2 = [0, 0, 20]
z1 = Reshape(x, s2)

Without the same root constraint, let’s add two expressions to above example:
| Formula |                                                             Value|
|:-- |:-- |
|y = Ones(1, 1, 120)|y has shape (1, 1, 120)|
|z2 = reshape(y, s2)|z has shape (2, 3, 20)|

We cannot optimize it as:
s2 = [0, 0, 20]
z2 = Reshape(y, s2)
It means to reshape to (1, 1, 20), which will fail since y has 120 elements.

The fusion is enabled for all ONNX models by default. We need make sure  it could work on any graph besides GPT2.

Reply:
I've asked @liuziyue to help implement a case that has more strict constraint (from same root), but could help GPT2 in some degree:

![image](https://user-images.githubusercontent.com/30328909/81745389-9d489000-9459-11ea-89e8-186c38a0bd0d.png)